File: backend/app/api/routes/lookups.py
from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.security.deps import require_roles
from app.schemas.lookup import (
    AssetModelCreateRequest,
    AssetModelUpdateRequest,
    LookupCreateRequest,
    LookupItemResponse,
    LookupListResponse,
    LookupUpdateRequest,
    VendorCreateRequest,
    VendorUpdateRequest,
)
from app.services.lookup import LookupService

router = APIRouter()


@router.get("/departments", response_model=LookupListResponse)
def list_departments(
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER", "OPERATOR", "VIEWER")),
) -> LookupListResponse:
    return LookupService(db).list_departments()


@router.post("/departments", response_model=LookupItemResponse)
def create_department(
    payload: LookupCreateRequest,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> LookupItemResponse:
    return LookupService(db).create_department(payload)


@router.put("/departments/{department_id}", response_model=LookupItemResponse)
def update_department(
    department_id: int,
    payload: LookupUpdateRequest,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> LookupItemResponse:
    return LookupService(db).update_department(department_id, payload)


@router.delete("/departments/{department_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_department(
    department_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> Response:
    LookupService(db).delete_department(department_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/locations", response_model=LookupListResponse)
def list_locations(
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER", "OPERATOR", "VIEWER")),
) -> LookupListResponse:
    return LookupService(db).list_locations()


@router.post("/locations", response_model=LookupItemResponse)
def create_location(
    payload: LookupCreateRequest,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> LookupItemResponse:
    return LookupService(db).create_location(payload)


@router.put("/locations/{location_id}", response_model=LookupItemResponse)
def update_location(
    location_id: int,
    payload: LookupUpdateRequest,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> LookupItemResponse:
    return LookupService(db).update_location(location_id, payload)


@router.delete("/locations/{location_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_location(
    location_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> Response:
    LookupService(db).delete_location(location_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/vendors", response_model=LookupListResponse)
def list_vendors(
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER", "OPERATOR", "VIEWER")),
) -> LookupListResponse:
    return LookupService(db).list_vendors()


@router.post("/vendors", response_model=LookupItemResponse)
def create_vendor(
    payload: VendorCreateRequest,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> LookupItemResponse:
    return LookupService(db).create_vendor(payload)


@router.put("/vendors/{vendor_id}", response_model=LookupItemResponse)
def update_vendor(
    vendor_id: int,
    payload: VendorUpdateRequest,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> LookupItemResponse:
    return LookupService(db).update_vendor(vendor_id, payload)


@router.delete("/vendors/{vendor_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_vendor(
    vendor_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> Response:
    LookupService(db).delete_vendor(vendor_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/asset-categories", response_model=LookupListResponse)
def list_asset_categories(
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER", "OPERATOR", "VIEWER")),
) -> LookupListResponse:
    return LookupService(db).list_asset_categories()


@router.post("/asset-categories", response_model=LookupItemResponse)
def create_asset_category(
    payload: LookupCreateRequest,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> LookupItemResponse:
    return LookupService(db).create_asset_category(payload)


@router.put("/asset-categories/{category_id}", response_model=LookupItemResponse)
def update_asset_category(
    category_id: int,
    payload: LookupUpdateRequest,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> LookupItemResponse:
    return LookupService(db).update_asset_category(category_id, payload)


@router.delete("/asset-categories/{category_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_asset_category(
    category_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> Response:
    LookupService(db).delete_asset_category(category_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/asset-statuses", response_model=LookupListResponse)
def list_asset_statuses(
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER", "OPERATOR", "VIEWER")),
) -> LookupListResponse:
    return LookupService(db).list_asset_statuses()


@router.get("/asset-models", response_model=LookupListResponse)
def list_asset_models(
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER", "OPERATOR", "VIEWER")),
) -> LookupListResponse:
    return LookupService(db).list_asset_models()


@router.post("/asset-models", response_model=LookupItemResponse)
def create_asset_model(
    payload: AssetModelCreateRequest,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> LookupItemResponse:
    return LookupService(db).create_asset_model(payload)


@router.put("/asset-models/{model_id}", response_model=LookupItemResponse)
def update_asset_model(
    model_id: int,
    payload: AssetModelUpdateRequest,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> LookupItemResponse:
    return LookupService(db).update_asset_model(model_id, payload)


@router.delete("/asset-models/{model_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_asset_model(
    model_id: int,
    db: Session = Depends(get_db),
    current_user=Depends(require_roles("ADMIN", "ASSET_MANAGER")),
) -> Response:
    LookupService(db).delete_asset_model(model_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/features/lookups/api/lookups.ts
import { apiRequest } from "@/api/http";
import type {
  AssetModelCreatePayload,
  AssetModelUpdatePayload,
  LookupCreatePayload,
  LookupListResponse,
  LookupReference,
  LookupUpdatePayload,
  VendorCreatePayload,
  VendorUpdatePayload,
} from "@/types/api";

export function getDepartments() {
  return apiRequest<LookupListResponse>("/departments");
}

export function getLocations() {
  return apiRequest<LookupListResponse>("/locations");
}

export function getVendors() {
  return apiRequest<LookupListResponse>("/vendors");
}

export function getAssetCategories() {
  return apiRequest<LookupListResponse>("/asset-categories");
}

export function getAssetStatuses() {
  return apiRequest<LookupListResponse>("/asset-statuses");
}

export function getAssetModels() {
  return apiRequest<LookupListResponse>("/asset-models");
}

export function createDepartment(payload: LookupCreatePayload) {
  return apiRequest<LookupReference>("/departments", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateDepartment(id: number, payload: LookupUpdatePayload) {
  return apiRequest<LookupReference>(`/departments/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function deleteDepartment(id: number) {
  return apiRequest<void>(`/departments/${id}`, {
    method: "DELETE",
  });
}

export function createLocation(payload: LookupCreatePayload) {
  return apiRequest<LookupReference>("/locations", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateLocation(id: number, payload: LookupUpdatePayload) {
  return apiRequest<LookupReference>(`/locations/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function deleteLocation(id: number) {
  return apiRequest<void>(`/locations/${id}`, {
    method: "DELETE",
  });
}

export function createVendor(payload: VendorCreatePayload) {
  return apiRequest<LookupReference>("/vendors", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateVendor(id: number, payload: VendorUpdatePayload) {
  return apiRequest<LookupReference>(`/vendors/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function deleteVendor(id: number) {
  return apiRequest<void>(`/vendors/${id}`, {
    method: "DELETE",
  });
}

export function createAssetCategory(payload: LookupCreatePayload) {
  return apiRequest<LookupReference>("/asset-categories", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateAssetCategory(id: number, payload: LookupUpdatePayload) {
  return apiRequest<LookupReference>(`/asset-categories/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function deleteAssetCategory(id: number) {
  return apiRequest<void>(`/asset-categories/${id}`, {
    method: "DELETE",
  });
}

export function createAssetModel(payload: AssetModelCreatePayload) {
  return apiRequest<LookupReference>("/asset-models", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateAssetModel(id: number, payload: AssetModelUpdatePayload) {
  return apiRequest<LookupReference>(`/asset-models/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function deleteAssetModel(id: number) {
  return apiRequest<void>(`/asset-models/${id}`, {
    method: "DELETE",
  });
}
